crates/cli-support/src/js/js2rust.rs

```diff
@@ -790,7 +810,11 @@ impl<'a, 'b> Js2Rust<'a, 'b> {
         let ts_args = self
             .js_arguments
             .iter()
-            .map(|s| format!("{}: {}", s.0, s.1))
+            .map(|s| if s.optional {
+                format!("{}?: {}", s.name, s.type_)
+            } else {
+                format!("{}: {}", s.name, s.type_)
+            })
```

Comment:
I considered that optional parameter is available or not. But my conclusion was 'available but not easy'. `foo: T | undefined` and `foo?: T` are not the same in terms of the point that the parameter can be omitted.

For example, TypeScript does not allow following function.

```typescript
function foo(a?: number, b: boolean) {}
```

Here is an error output:

```
foo.ts:1:26 - error TS1016: A required parameter cannot follow an optional parameter.

1 function foo(a?: number, b: boolean) {}
                           ~


Found 1 error.
```

So the parameter which can be optional is 'a parameter which is not followed by required parameter'.

For this reason, I used `T | undefined` always.

Reply:
Oh, that is a very good point... I guess it's complicated :confused: 

One possibility would be to add a wasm-bindgen wrapper type called `Omittable` which results in an optional field (and an error when a non-omittable field follows an omittable field). I'm not sure whether I like that though.

Another option would be to add a function attribute called `#[optional_omittable]` or something like that. If set, then all optional fields that are not followed by non-optional fields would be marked as omittable (using `?`).

Reply:
I wrote 'not easy' but if we only use optional parameter for parameters not followed by required parameters, I think following function would be enough.

```rust
let mut omittable = true;
let mut ts_args = vec![];
for arg in self.js_arguments.iter().rev() {
    if arg.optional && omittable {
        if omittable {
            ts_args.push(format!("{}?: {}", name, type_)
        } else {
            ts_args.push(format!("{}: {} | undefined", name, type_)
        }
    } else {
        omittable = false;
        ts_args.push(format!("{}: {}", name, type_)
    }
}
ts_args.reverse();
```

Reply:
Thank you @rhysd, I integrated this logic into the PR!